src/consensus/aft/raft.h

```diff
@@ -2515,19 +2676,61 @@ namespace aft
 
     void become_retired()
     {
-      replica_state = kv::ReplicaState::Retired;
-      leader_id.reset();
+      catching_up = false;
 
-      LOG_INFO_FMT(
-        "Becoming retired {}: {}", state->my_node_id, state->current_view);
+#if 0
+      // To guarantee liveness during reconfiguration, we need something of this
+      // sort (e.g. when replacing the whole network in one reconfiguration).
+      // This is not enough though; we would need to check that n/2+1 of the new
+      // nodes have seen enough of the new configuration to make sure we don't
+      // lose any transactions.
```

Comment:
I agree with the comment, but I don't think a local decision to delay retirement will work. My understanding is; when a node knows locally of its own removal, it should stop producing writes but otherwise continues as a normal follower/primary (it is valid and correct for a node to become leader after its removal, and share this suffix describing its own removal with the other nodes). Once this is globally committed, the node can transition to `Retired` and cease to participate (with a slight complication that if it is primary, it _should_ stick a round a little longer to try to catch up _all_ nodes, and do what it can to ensure the next primary is live in the latest configuration).

I'm not sure how this changes with catch-up. I think if a node hears global commit of its removal, it can still retire instantly - this means there are a quorum of nodes in the new configuration who can take over, so they don't need this specific node to stay active any longer.